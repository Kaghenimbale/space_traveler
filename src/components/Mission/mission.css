.mission-content {
  padding: 1rem 0;
  width: 70vw;
  margin: 0 auto;
}

.table-mission {
  border: 1px solid darkgray;
  border-spacing: 0;
}

.loader {
  width: 100vw;
  display: flex;
  align-items: center;
  justify-content: center;
}

.mission th {
  border: 0.3px solid lightgray;
  padding: 1rem 2rem;
  font-weight: bolder;
}

tr td {
  border: 0.3px solid lightgray;
  padding: 1rem 2rem;
}

.title {
  font-weight: bolder;
  white-space: nowrap;
}

.table-item:nth-child(odd) {
  background-color: whitesmoke;
}

.table-btn {
  white-space: nowrap;
  padding: 0.3rem 0.6rem;
}

.status p {
  white-space: nowrap;
  background: darkgray;
  color: white;
  padding: 0.3rem;
}
