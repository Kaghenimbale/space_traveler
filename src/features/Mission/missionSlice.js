import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const initialState = {
  missionItems: [],
  isLoading: false,
};

const url = 'https://api.spacexdata.com/v3/missions';

export const fetchMission = createAsyncThunk(
  'rockect/fecthMission',
  async () => {
    try {
      const response = await axios.get(url);
      return response.data;
    } catch (error) {
      return error.name;
    }
  },
);

const missionSlice = createSlice({
  name: 'mission',
  initialState,
  extraReducers(builder) {
    builder
      .addCase(fetchMission.pending, (state) => ({
        ...state,
        isLoading: true,
      }))
      .addCase(fetchMission.fulfilled, (state, action) => ({
        ...state,
        isLoading: false,
        missionItems: action.payload,
      }))
      .addCase(fetchMission.rejected, (state) => ({
        ...state,
        isLoading: false,
      }));
  },
});

export default missionSlice.reducer;
