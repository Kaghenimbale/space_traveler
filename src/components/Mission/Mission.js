import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { MoonLoader } from 'react-spinners';
import {
  fetchMission,
  reserveMission,
} from '../../features/Mission/missionSlice';
import './mission.css';

const Mission = () => {
  const {
    missionItems, isLoading, error, isFetched,
  } = useSelector(
    (state) => state.mission,
  );

  const dispatch = useDispatch();

  useEffect(() => {
    if (!isFetched) {
      dispatch(fetchMission());
    }
  }, [dispatch, isFetched]);

  const handleReserved = (id) => {
    dispatch(reserveMission({ id }));
  };

  if (isLoading) {
    return (
      <div className="loader">
        <MoonLoader />
      </div>
    );
  }

  if (error) {
    return <h1>Something went Wrong!</h1>;
  }

  return (
    <div className="mission-content">
      <table className="table-mission">
        <thead>
          <tr className="mission">
            <th>Mission</th>
            <th>Description</th>
            <th>Status</th>
          </tr>
        </thead>
        <tbody className="table-body">
          {missionItems.map((item) => (
            <tr className="table-item" key={item.mission_id}>
              <td className="title">{item.mission_name}</td>
              <td className="description">{item.description}</td>
              <td className="status">
                <p
                  className={
                    !item.reserved ? 'statusNotActive' : 'statusActive'
                  }
                >
                  {!item.reserved ? 'NOT A MEMBER' : 'Active Member'}
                </p>
              </td>
              <td>
                <button
                  className={!item.reserved ? 'table-btn' : 'tableBtnActive'}
                  type="button"
                  onClick={() => handleReserved(item.mission_id)}
                >
                  {!item.reserved ? 'Join Mission' : 'Leave Mission'}
                </button>
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

export default Mission;
