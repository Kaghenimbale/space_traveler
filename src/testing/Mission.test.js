import { render, cleanup, screen } from '@testing-library/react';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import Mission from '../components/Mission/Mission';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';

const mockStore = configureMockStore();
const store = mockStore({
  mission: {
    missionItems: [
      {
        mission_id: '1',
        mission_name: 'Go to Mars',
        description: 'Mars is a great planet',
        reserved: false,
      },
      {
        mission_id: '2',
        mission_name: 'Go to Jupiter',
        description: 'Jupiter is a great planet',
        reserved: false,
      },
      {
        mission_id: '3',
        mission_name: 'Go to Saturn',
        description: 'Saturn is a great planet',
        reserved: false,
      },
    ],
    isFetched: true,
  },
});
afterEach(cleanup);

describe('Render UI', () => {
  it('Render Component', () => {
    render(
      <Provider store={store}>
        <Mission />
      </Provider>,
    );
    expect(screen.getByText('Go to Saturn')).toBeInTheDocument();
  });
  it('Render mission lists', () => {
    const { container } = render(
      <Provider store={store}>
        <Mission />
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
