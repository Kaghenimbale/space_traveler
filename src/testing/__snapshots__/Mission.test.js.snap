// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render UI Render mission lists 1`] = `
<div>
  <div
    class="mission-content"
  >
    <table
      class="table-mission"
    >
      <thead>
        <tr
          class="mission"
        >
          <th>
            Mission
          </th>
          <th>
            Description
          </th>
          <th>
            Status
          </th>
        </tr>
      </thead>
      <tbody
        class="table-body"
      >
        <tr
          class="table-item"
        >
          <td
            class="title"
          >
            Go to Mars
          </td>
          <td
            class="description"
          >
            Mars is a great planet
          </td>
          <td
            class="status"
          >
            <p
              class="statusNotActive"
            >
              NOT A MEMBER
            </p>
          </td>
          <td>
            <button
              class="table-btn"
              type="button"
            >
              Join Mission
            </button>
          </td>
        </tr>
        <tr
          class="table-item"
        >
          <td
            class="title"
          >
            Go to Jupiter
          </td>
          <td
            class="description"
          >
            Jupiter is a great planet
          </td>
          <td
            class="status"
          >
            <p
              class="statusNotActive"
            >
              NOT A MEMBER
            </p>
          </td>
          <td>
            <button
              class="table-btn"
              type="button"
            >
              Join Mission
            </button>
          </td>
        </tr>
        <tr
          class="table-item"
        >
          <td
            class="title"
          >
            Go to Saturn
          </td>
          <td
            class="description"
          >
            Saturn is a great planet
          </td>
          <td
            class="status"
          >
            <p
              class="statusNotActive"
            >
              NOT A MEMBER
            </p>
          </td>
          <td>
            <button
              class="table-btn"
              type="button"
            >
              Join Mission
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
